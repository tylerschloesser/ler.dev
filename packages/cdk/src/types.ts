import { Environment, StackProps } from 'aws-cdk-lib'

export enum Stage {
  Prod = 'prod',
  Staging = 'staging',
}

export enum Region {
  US_WEST_2 = 'us-west-2',
  US_EAST_1 = 'us-east-1',
}

export interface CommonStackProps extends StackProps {
  env: Omit<Required<Environment>, 'region'> & {
    region: Region
  }
}

export enum DomainName {
  LerDev = 'ler.dev',
  TyLerDev = 'ty.ler.dev',
  StagingTyLerDev = 'staging.ty.ler.dev',
  DrawApiTyLerDev = 'draw-api.ty.ler.dev',
  DrawApiStagingTyLerDev = 'draw-api.staging.ty.ler.dev',
}
