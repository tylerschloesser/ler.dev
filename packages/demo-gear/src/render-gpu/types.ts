import { mat4 } from 'gl-matrix'

export type ShaderType = number
export type ShaderSource = string
export type WebGLAttributeLocation = number

export interface GpuState {
  programs: {
    grid: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        viewport: WebGLUniformLocation | null
        tileSize: WebGLUniformLocation | null
        camera: WebGLUniformLocation | null
        pixelRatio: WebGLUniformLocation | null
        zoom: WebGLUniformLocation | null
      }
    }
    main: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        model: WebGLUniformLocation
        view: WebGLUniformLocation
        projection: WebGLUniformLocation
        color: WebGLUniformLocation
      }
    }
  }
  buffers: {
    square: WebGLBuffer
  }
  textures: {
    gears: Record<number, WebGLTexture>
  }
  matrices: {
    model: mat4
    view: mat4
    projection: mat4
  }
}
