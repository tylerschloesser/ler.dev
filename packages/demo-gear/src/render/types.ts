import { mat4 } from 'gl-matrix'

export type ShaderType = number
export type ShaderSource = string
export type WebGLAttributeLocation = number

export interface GpuState {
  programs: {
    grid: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        viewport: WebGLUniformLocation | null
        tileSize: WebGLUniformLocation | null
        camera: WebGLUniformLocation | null
        zoom: WebGLUniformLocation | null
      }
    }
    gearBody: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        model: WebGLUniformLocation
        view: WebGLUniformLocation
        projection: WebGLUniformLocation
        color: WebGLUniformLocation
      }
    }
    gearTeeth: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        model: WebGLUniformLocation
        view: WebGLUniformLocation
        projection: WebGLUniformLocation
        color: WebGLUniformLocation
        tileSize: WebGLUniformLocation
      }
    }
    chain: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        model: WebGLUniformLocation
        view: WebGLUniformLocation
        projection: WebGLUniformLocation
        color: WebGLUniformLocation
      }
    }
    outlineRect: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        model: WebGLUniformLocation
        view: WebGLUniformLocation
        projection: WebGLUniformLocation
        color: WebGLUniformLocation
        size: WebGLUniformLocation
      }
    }
    fillRect: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
      }
      uniforms: {
        model: WebGLUniformLocation
        view: WebGLUniformLocation
        projection: WebGLUniformLocation
        color: WebGLUniformLocation
      }
    }
    fillInstanced: {
      program: WebGLProgram
      attributes: {
        vertex: WebGLAttributeLocation
        matrix: WebGLAttributeLocation
      }
      uniforms: {
        // view: WebGLUniformLocation
        // projection: WebGLUniformLocation
        transform: WebGLUniformLocation
        color: WebGLUniformLocation
      }
    }
  }
  buffers: {
    square: WebGLBuffer
    fillRect: WebGLBuffer
    outlineRect: WebGLBuffer
    gearBody: Record<
      number,
      {
        vertex: WebGLBuffer
        count: number
      }
    >
    gearTooth: WebGLBuffer
    fillInstancedMatrices: {
      data: Float32Array
      buffer: WebGLBuffer
      values: mat4[]
    }
  }
  matrices: {
    model: mat4
    view: mat4
    projection: mat4
  }
}
