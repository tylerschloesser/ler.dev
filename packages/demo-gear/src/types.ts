import * as z from 'zod'

export const SimpleVec2 = z.strictObject({
  x: z.number(),
  y: z.number(),
})
export type SimpleVec2 = z.infer<typeof SimpleVec2>

export const GearId = z.string()
export type GearId = z.infer<typeof GearId>

export const TileId = z.string()
export type TileId = z.infer<typeof TileId>

export const Tile = z.strictObject({
  gearId: GearId,
  attachedGearId: GearId.optional(),
})
export type Tile = z.infer<typeof Tile>

export const ConnectionType = z.enum([
  'Adjacent',
  'Chain',
  'Attach',
])
export type ConnectionType = z.infer<typeof ConnectionType>

export const Connection = z.strictObject({
  type: ConnectionType,
  gearId: GearId,
})
export type Connection = z.infer<typeof Connection>

export const Gear = z.strictObject({
  id: GearId,
  position: SimpleVec2,
  radius: z.number(),
  angle: z.number(),
  velocity: z.number(),
  mass: z.number(),
  connections: z.array(Connection),
})
export type Gear = z.infer<typeof Gear>

export const World = z.strictObject({
  gears: z.record(GearId, Gear),
  tiles: z.record(TileId, Tile),

  debugConnections: z.boolean(),
})
export type World = z.infer<typeof World>

export enum HandType {
  Build = 'build',
  Accelerate = 'accelerate',
}

interface BaseHand<T extends HandType> {
  type: T
  position: SimpleVec2 | null
}

export interface BuildHand
  extends BaseHand<HandType.Build> {
  radius: number
  valid: boolean
  chain: Gear | null
  connections: Connection[]
  onChangeValid?(valid: boolean): void
}

export interface AccelerateHand
  extends BaseHand<HandType.Accelerate> {
  active: boolean
  direction: number
  gear: Gear | null
  onChangeGear?(gear: Gear | null): void
}

export type Hand = BuildHand | AccelerateHand

export type SetWorldFn = (world: World) => void

export interface Camera {
  position: SimpleVec2
  zoom: number
}

export type PointerListenerFn = (
  state: AppState,
  e: PointerEvent,
  position: Readonly<SimpleVec2>,
) => void
export type CameraListenerFn = (state: AppState) => void

type Texture = Blob

export interface Textures {
  gears: Record<number, Texture>
}

export interface AppState {
  canvas: {
    container: HTMLDivElement
    cpu: HTMLCanvasElement
    gpu: HTMLCanvasElement
  }
  viewport: {
    size: SimpleVec2
    pixelRatio: number
  }
  signal: AbortSignal
  world: World
  setWorld: SetWorldFn
  hand: Hand | null

  camera: Camera
  tileSize: number

  pointerListeners: Set<PointerListenerFn>
  cameraListeners: Set<CameraListenerFn>

  textures: Textures
}

export type InitFn = (state: AppState) => Promise<void>

export type PartialGear = Pick<
  Gear,
  'position' | 'radius' | 'angle'
>
