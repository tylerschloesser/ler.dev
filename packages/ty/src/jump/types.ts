import { Viewport } from '../common/engine'
import { Vec2 } from '../common/vec2'

export type TargetIndex = number

export interface Ball {
  p: Vec2
  v: Vec2
  r: number

  capturedBy: TargetIndex | null
  launchedBy: TargetIndex | null
}

export interface Target {
  p: Vec2
  r: number
}

export interface Drag {
  a: Vec2
  b: Vec2 | null
}

export interface Camera {
  p: Vec2
  v: Vec2
}

export interface State {
  pointer: Vec2 | null
  drag: Drag | null
  ball: Ball | null
  viewport: Viewport
  targets: Target[]
  camera: Camera
}
