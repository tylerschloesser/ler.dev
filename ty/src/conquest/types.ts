import Color from 'color'
import { Vec2 } from '../common/vec2'

export interface Flag {
  p: Vec2
  r: number
  color: Color
  progress: number
}

export interface Ball {
  p: Vec2
  v: Vec2
  r: number
  color: Color
}

export interface World {
  size: Vec2
  flags: Flag[]
}

export type Pointer = Vec2 | null

export type Drag = null | {
  a: Vec2
  b: Vec2 | null
}

export interface Camera {
  p: Vec2
  zoom: number
}

export interface State {
  pointer: Pointer
  drag: Drag
  world: World
  camera: Camera
  ball: Ball

  closestFlagInfo: null | {
    index: number
    modifier: Vec2
  }
}
