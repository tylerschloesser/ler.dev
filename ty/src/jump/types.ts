import { Vec2 } from '../common/vec2'

export interface Ball {
  p: Vec2
  v: Vec2
}

export interface Drag {
  a: Vec2
  b: Vec2 | null
}

export interface State {
  pointer: Vec2 | null
  drag: Drag | null
  ball: Ball | null
}
