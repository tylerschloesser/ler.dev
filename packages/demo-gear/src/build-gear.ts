import invariant from 'tiny-invariant'
import {
  ConnectionType,
  EntityType,
  GearEntity,
  World,
} from './types.js'
import { getEntity, iterateGearTileIds } from './util.js'

export function buildGear(
  world: World,
  gear: GearEntity,
): void {
  let attach: GearEntity | null = null
  for (const connection of gear.connections) {
    if (connection.type === ConnectionType.enum.Attach) {
      const entity = getEntity(world, connection.entityId)
      invariant(entity.type === EntityType.enum.Gear)
      invariant(attach === null)
      attach = entity
    }
  }

  invariant(world.entities[gear.id] === undefined)

  world.entities[gear.id] = gear

  for (const tileId of iterateGearTileIds(
    gear.center,
    gear.radius,
  )) {
    let tile = world.tiles[tileId]

    if (attach) {
      invariant(tile?.entityId === attach.id)
      tile.entityId = gear.id
    } else {
      invariant(tile === undefined)
      tile = world.tiles[tileId] = { entityId: gear.id }
    }
  }
}
