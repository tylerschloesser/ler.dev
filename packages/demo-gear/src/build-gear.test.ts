import { describe, expect, test } from '@jest/globals'
import { getBuildHand } from './build-gear.js'
import { SimpleVec2, World } from './types.js'

describe('build-gear', () => {
  describe('getBuildHand', () => {
    test('base', () => {
      const world: World = {
        entities: {},
        networks: {},
        tiles: {},
        version: 1,
      }
      const center: SimpleVec2 = {
        x: 0,
        y: 0,
      }
      const radius = 1
      const chainFrom = null
      const hand = getBuildHand(
        world,
        center,
        radius,
        chainFrom,
      )
      expect(hand).toMatchSnapshot()
    })
  })
})
