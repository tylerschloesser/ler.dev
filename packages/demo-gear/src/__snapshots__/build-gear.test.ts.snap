// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-gear getBuildHand base 1`] = `
{
  "action": {
    "type": "build",
  },
  "entities": {
    "-1.-1": {
      "angle": 0,
      "center": {
        "x": 0,
        "y": 0,
      },
      "connections": [],
      "id": "-1.-1",
      "mass": 3.141592653589793,
      "networkId": "-1.-1",
      "position": {
        "x": -1,
        "y": -1,
      },
      "radius": 1,
      "size": {
        "x": 2,
        "y": 2,
      },
      "type": "Gear",
      "velocity": 0,
    },
  },
  "networks": {
    "-1.-1": {
      "entityIds": {
        "-1.-1": true,
      },
      "id": "-1.-1",
      "mass": 3.141592653589793,
      "rootId": "-1.-1",
    },
  },
  "type": "build",
  "valid": true,
}
`;
